lib/OrdinaryDiffEqNonlinearSolve/src/newton.jl

```diff
@@ -63,13 +63,29 @@ function initialize!(nlsolver::NLSolver{<:NonlinearSolveAlg, true},
         integrator.stats.nnonliniter += cache.cache.stats.nsteps
         integrator.stats.njacs += cache.cache.stats.njacs
     end
-    if f isa DAEFunction
-        nlp_params = (tmp, ztmp, ustep, γ, α, tstep, k, invγdt, p, dt, f)
+
+    nlstep_data = f.nlstep_data
+    if nlstep_data !== nothing
+        if method === COEFFICIENT_MULTISTEP
+            nlstep_data.set_γ_c(nlstep_data.nlprob, (one(t), one(t), α * invγdt, tstep))
+            nlstep_data.set_inner_tmp(nlstep_data.nlprob, zero(z))
+            nlstep_data.set_outer_tmp(nlstep_data.nlprob, tmp)
+        else
+            nlstep_data.set_γ_c(nlstep_data.nlprob, (dt, γ, one(t)))
+            nlstep_data.set_inner_tmp(nlstep_data.nlprob, tmp)
+            nlstep_data.set_outer_tmp(nlstep_data.nlprob, zero(z))
+        end
+        nlstep_data.nlprob.u0 .= @view z[nlstep_data.u0perm]
+        cache.cache = init(nlstep_data.nlprob, alg.alg)
```

Comment:
One performance enhancement could be to test whether updating `u0` and `p` directly on the cache works as expected. It may just need to clear memory but it should work.